File: cpp/bob/bob.cpp
#include "bob.h"
#include <algorithm>

namespace bob
{
    std::string hey(std::string input)
    {
        if (std::all_of(input.begin(), input.end(), [](char c)
                        { if (isalpha(c)) return std::isupper(c); return -1; }) and
            input[input.length() - 1] == '?')
        {
            return "Calm down, I know what I'm doing!";
        }
        if (std::all_of(input.begin(), input.end(), [](char c)
                        { if (isalpha(c)) return std::isupper(c); return -1; }))
        {
            return "Whoa, chill out!";
        }
        if (std::all_of(input.begin(), input.end(), [](char c)
                        { return std::isspace(c); }))
        {
            return "Fine. Be that way!";
        }
        if (input[input.length() - 1] == '?')
        {
            return "Sure.";
        }
        return "Whatever.";
    }
} // namespace bob

File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"strings"
	"unicode"
)

var REPLACE_LIST = []string{
	" ", "\t", "\n", "\r",
}

// Hey should have a comment documenting it.
func Hey(remark string) string {
	s := Cleanup(remark)
	if s == "" {
		return "Fine. Be that way!"
	}

	isQuestion := strings.HasSuffix(s, "?")
	isAllUpper := isAllUpper(s)

	if isQuestion && isAllUpper {
		return "Calm down, I know what I'm doing!"
	}

	if isQuestion {
		return "Sure."
	}

	if isAllUpper {
		return "Whoa, chill out!"
	}

	return "Whatever."
}

func Cleanup(s string) string {
	for _, v := range REPLACE_LIST {
		s = strings.ReplaceAll(s, string(v), "")
	}
	return s
}

func isAllUpper(s string) bool {
	hasLetters := false
	allUpper := true
	for _, v := range s {
		if unicode.IsLetter(v) {
			hasLetters = true
			if !unicode.IsUpper(v) {
				allUpper = false
			}
		}
	}
	return hasLetters && allUpper
}
